lumicks/pylake/tests/test_channels.py

```diff
@@ -576,3 +577,8 @@ def testLine(x, y):
     mpl.pyplot.gca().clear()
     s.plot(start=100e9)
     testLine(np.arange(0, 230, 10) - 100 + 5, d)
+
+
+def test_regression_lazy_loading(h5_file):
+    ch = channel.Continuous.from_dataset(h5_file["Force HF"]["Force 1x"])
+    assert type(ch._src._src_data) == h5py.Dataset
```

Comment:
Would it make sense to add some sort of progress report to the tests in general? I noticed running these succeed but there is no terminal output like `Test blah blah passed`, which probably means an error would be a scary dump to the terminal.

Reply:
Hmm. That's strange. There should be terminal input in the following form already:
```python
============================= test session starts ==============================
platform linux -- Python 3.6.12, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
rootdir: /home/runner/work/pylake/pylake, configfile: setup.cfg
collected 304 items

lumicks/pylake/fitting/tests/test_profiles.py ...                        [  0%]
lumicks/pylake/fitting/tests/test_stepping.py .......................... [  9%]
.....                                                                    [ 11%]
lumicks/pylake/force_calibration/tests/test_power_spectrum.py .......... [ 14%]
.........                                                                [ 17%]
lumicks/pylake/force_calibration/tests/test_power_spectrum_calibration.py . [ 17%]
..........................                                               [ 26%]
lumicks/pylake/kymotracker/tests/test_io.py ..........                   [ 29%]
lumicks/pylake/kymotracker/tests/test_peakfinding.py .......             [ 31%]
lumicks/pylake/kymotracker/tests/test_refinement.py ......               [ 33%]
lumicks/pylake/kymotracker/tests/test_tracing.py ......                  [ 35%]
lumicks/pylake/nb_widgets/tests/test_fd_selector.py .....                [ 37%]
lumicks/pylake/nb_widgets/tests/test_kymotracker_widget.py ......        [ 39%]
lumicks/pylake/nb_widgets/tests/test_mouse.py ..                         [ 40%]
lumicks/pylake/tests/test_channels.py .............                      [ 44%]
lumicks/pylake/tests/test_dfcurve.py .                                   [ 44%]
lumicks/pylake/tests/test_file.py .............                          [ 49%]
lumicks/pylake/tests/test_kymo.py .......                                [ 51%]
lumicks/pylake/tests/test_point_scan.py ..                               [ 51%]
lumicks/pylake/tests/test_scan.py .....                                  [ 53%]
lumicks/pylake/tests/test_channels.py ...                                [ 54%]
lumicks/pylake/tests/test_dfcurve.py .                                   [ 54%]
lumicks/pylake/tests/test_file.py .............                          [ 59%]
lumicks/pylake/tests/test_kymo.py .......                                [ 61%]
lumicks/pylake/tests/test_point_scan.py ..                               [ 62%]
lumicks/pylake/tests/test_scan.py .....                                  [ 63%]
lumicks/pylake/tests/test_channels.py ...........                        [ 67%]
lumicks/pylake/tests/test_correlated_stack.py ..........                 [ 70%]
lumicks/pylake/tests/test_dfcurve.py ............                        [ 74%]
lumicks/pylake/tests/test_fdensemble.py .......                          [ 76%]
lumicks/pylake/tests/test_fdfit.py ...........................           [ 85%]
lumicks/pylake/tests/test_file.py ...                                    [ 86%]
lumicks/pylake/tests/test_image.py .......                               [ 89%]
lumicks/pylake/tests/test_import_time.py .                               [ 89%]
lumicks/pylake/tests/test_kymotracker.py ......................          [ 96%]
lumicks/pylake/tests/test_mixin.py ...                                   [ 97%]
lumicks/pylake/tests/test_timeindex.py ...                               [ 98%]
lumicks/pylake/tests/test_utilities.py ....                              [100%]

============================= 304 passed in 12.10s =============================
```

Pytest handles the printing. If a test fails, you get a print which says which test failed including a stack trace. The errors usually aren't too bad. You'd get something like this:
```
=================================== FAILURES ===================================
__________________________ test_confidence_intervals ___________________________

    def test_confidence_intervals():
        eps = 1e-5
        x = np.arange(7)
    
        # The following is generated from y = 4.0 * x + 5.0 + 2.0*np.random.randn(N):
        y = [8.24869073,  7.77648717, 11.9436565, 14.85406276, 22.73081526, 20.39692261, 32.48962353]
    
        def validate_profile(name, func, jac, x, y, parameter, stderr, lb, ub):
            F = Fit(Model(name, func, jacobian=jac))
            F._add_data(name, x, y)
            F.fit()
            profile = F.profile_likelihood(name + "/" + parameter)
    
            assert (F[name + "/" + parameter].stderr - stderr) < eps
            assert F[name + "/" + parameter].profile == profile
            assert abs(profile.lower_bound - lb) < eps
            assert abs(profile.upper_bound - ub) < eps
            assert len(profile.chi2) == len(profile.p)
    
            # Validate asymptotic confidence intervals based on comparing them with the profiles
            assert abs(F[name + "/" + parameter].ci(.95)[0] - profile.lower_bound) < 1e-2
            assert abs(F[name + "/" + parameter].ci(.95)[1] - profile.upper_bound) < 1e-2
    
>       validate_profile("linear", linear, linear_jac, x, y, "a", 0.61939832586, 2.6713088215948138, 5.096607466976615)

lumicks/pylake/fitting/tests/test_profiles.py:58: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
lumicks/pylake/fitting/tests/test_profiles.py:45: in validate_profile
    F.fit()
lumicks/pylake/fitting/fit.py:217: in fit
    parameter_vector, fitted, lb, ub = self._prepare_fit()
lumicks/pylake/fitting/fit.py:159: in _prepare_fit
    return self.params.values, self.params.fitted, self.params.lower_bounds, self.params.upper_bounds
lumicks/pylake/fitting/parameters.py:199: in fitted
    return np.asarray([not param.fixed for param in self._src.values()], dtype=np.bool)
```

I'd say the failed ones are the only ones you're interested in and the stack trace is probably the most informative thing for fixing the problem. Showing which ones pass would only produce a lot of noise.

Reply:
> Hmm. That's strange. There should be terminal input in the following form already:
> 
> ```python
> ============================= test session starts ==============================
> platform linux -- Python 3.6.12, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
> rootdir: /home/runner/work/pylake/pylake, configfile: setup.cfg
> collected 304 items
> 
> lumicks/pylake/fitting/tests/test_profiles.py ...                        [  0%]
> lumicks/pylake/fitting/tests/test_stepping.py .......................... [  9%]
> .....                                                                    [ 11%]
> lumicks/pylake/force_calibration/tests/test_power_spectrum.py .......... [ 14%]
> .........                                                                [ 17%]
> lumicks/pylake/force_calibration/tests/test_power_spectrum_calibration.py . [ 17%]
> ..........................                                               [ 26%]
> lumicks/pylake/kymotracker/tests/test_io.py ..........                   [ 29%]
> lumicks/pylake/kymotracker/tests/test_peakfinding.py .......             [ 31%]
> lumicks/pylake/kymotracker/tests/test_refinement.py ......               [ 33%]
> lumicks/pylake/kymotracker/tests/test_tracing.py ......                  [ 35%]
> lumicks/pylake/nb_widgets/tests/test_fd_selector.py .....                [ 37%]
> lumicks/pylake/nb_widgets/tests/test_kymotracker_widget.py ......        [ 39%]
> lumicks/pylake/nb_widgets/tests/test_mouse.py ..                         [ 40%]
> lumicks/pylake/tests/test_channels.py .............                      [ 44%]
> lumicks/pylake/tests/test_dfcurve.py .                                   [ 44%]
> lumicks/pylake/tests/test_file.py .............                          [ 49%]
> lumicks/pylake/tests/test_kymo.py .......                                [ 51%]
> lumicks/pylake/tests/test_point_scan.py ..                               [ 51%]
> lumicks/pylake/tests/test_scan.py .....                                  [ 53%]
> lumicks/pylake/tests/test_channels.py ...                                [ 54%]
> lumicks/pylake/tests/test_dfcurve.py .                                   [ 54%]
> lumicks/pylake/tests/test_file.py .............                          [ 59%]
> lumicks/pylake/tests/test_kymo.py .......                                [ 61%]
> lumicks/pylake/tests/test_point_scan.py ..                               [ 62%]
> lumicks/pylake/tests/test_scan.py .....                                  [ 63%]
> lumicks/pylake/tests/test_channels.py ...........                        [ 67%]
> lumicks/pylake/tests/test_correlated_stack.py ..........                 [ 70%]
> lumicks/pylake/tests/test_dfcurve.py ............                        [ 74%]
> lumicks/pylake/tests/test_fdensemble.py .......                          [ 76%]
> lumicks/pylake/tests/test_fdfit.py ...........................           [ 85%]
> lumicks/pylake/tests/test_file.py ...                                    [ 86%]
> lumicks/pylake/tests/test_image.py .......                               [ 89%]
> lumicks/pylake/tests/test_import_time.py .                               [ 89%]
> lumicks/pylake/tests/test_kymotracker.py ......................          [ 96%]
> lumicks/pylake/tests/test_mixin.py ...                                   [ 97%]
> lumicks/pylake/tests/test_timeindex.py ...                               [ 98%]
> lumicks/pylake/tests/test_utilities.py ....                              [100%]
> 
> ============================= 304 passed in 12.10s =============================
> ```
> 
> Pytest handles the printing. If a test fails, you get a print which says which test failed including a stack trace. The errors usually aren't too bad. You'd get something like this:
> 
> ```
> =================================== FAILURES ===================================
> __________________________ test_confidence_intervals ___________________________
> 
>     def test_confidence_intervals():
>         eps = 1e-5
>         x = np.arange(7)
>     
>         # The following is generated from y = 4.0 * x + 5.0 + 2.0*np.random.randn(N):
>         y = [8.24869073,  7.77648717, 11.9436565, 14.85406276, 22.73081526, 20.39692261, 32.48962353]
>     
>         def validate_profile(name, func, jac, x, y, parameter, stderr, lb, ub):
>             F = Fit(Model(name, func, jacobian=jac))
>             F._add_data(name, x, y)
>             F.fit()
>             profile = F.profile_likelihood(name + "/" + parameter)
>     
>             assert (F[name + "/" + parameter].stderr - stderr) < eps
>             assert F[name + "/" + parameter].profile == profile
>             assert abs(profile.lower_bound - lb) < eps
>             assert abs(profile.upper_bound - ub) < eps
>             assert len(profile.chi2) == len(profile.p)
>     
>             # Validate asymptotic confidence intervals based on comparing them with the profiles
>             assert abs(F[name + "/" + parameter].ci(.95)[0] - profile.lower_bound) < 1e-2
>             assert abs(F[name + "/" + parameter].ci(.95)[1] - profile.upper_bound) < 1e-2
>     
> >       validate_profile("linear", linear, linear_jac, x, y, "a", 0.61939832586, 2.6713088215948138, 5.096607466976615)
> 
> lumicks/pylake/fitting/tests/test_profiles.py:58: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> lumicks/pylake/fitting/tests/test_profiles.py:45: in validate_profile
>     F.fit()
> lumicks/pylake/fitting/fit.py:217: in fit
>     parameter_vector, fitted, lb, ub = self._prepare_fit()
> lumicks/pylake/fitting/fit.py:159: in _prepare_fit
>     return self.params.values, self.params.fitted, self.params.lower_bounds, self.params.upper_bounds
> lumicks/pylake/fitting/parameters.py:199: in fitted
>     return np.asarray([not param.fixed for param in self._src.values()], dtype=np.bool)
> ```
> 
> I'd say the failed ones are the only ones you're interested in and the stack trace is probably the most informative thing for fixing the problem. Showing which ones pass would only produce a lot of noise.

You're right. I had only run it at the file level instead of the module level. This is totally fine 👍🏿